src/blockdata/block.rs

```diff
@@ -60,6 +60,62 @@ pub struct Block {
     pub txdata: Vec<Transaction>
 }
 
+impl Block {
+    /// check if merkle root of header matches merkle root of the transaction list
+    pub fn check_merkle_root (&self) -> bool {
+        self.header.merkle_root == self.merkle_root()
+    }
+
+    /// check if witness commitment in coinbase is matching the transaction list
+    pub fn check_witness_commitment(&self) -> bool {
+
+        // witness commitment is optional if there are no transactions using SegWit in the block
+        if self.txdata.iter().all(|t| t.input.iter().all(|i| i.witness.is_empty())) {
+            return true;
+        }
+        if self.txdata.len() > 0 {
+            let coinbase = &self.txdata[0];
+            if coinbase.is_coin_base() {
+                // commitment is in the last output that starts with below magic
+                if let Some(pos) = coinbase.output.iter()
+                    .rposition(|o| {
+                        o.script_pubkey.len () >= 38 &&
+                        o.script_pubkey[0..6] == [0x6a, 0x24, 0xaa, 0x21, 0xa9, 0xed] }) {
+                    let commitment = sha256d::Hash::from_slice(&coinbase.output[pos].script_pubkey.as_bytes()[6..38]).unwrap();
+                    // witness reserved value is in coinbase input witness
+                    if coinbase.input[0].witness.len() == 1 && coinbase.input[0].witness[0].len() == 32 {
+                        let witness_reserved_value = sha256d::Hash::from_slice(coinbase.input[0].witness[0].as_slice()).unwrap();
```

Comment:
The witness reserved value is just a 32 byte u8 I think.

Reply:
yes, it is. What is your point?

Reply:
Well I'm not sure why you're constructing a `sha256d::Hash` from it here

Reply:
You are right that the witness reserved value has no defined use yet. The BIP however states, that "The witness reserved value currently has no consensus meaning, but in the future allows new commitment values for future softforks." All commitments are sha256d hashes in Bitcoin (ids, merkle root, witness root) therefore I assumed that it will be also a sha256d hash.

Reply:
Even stronger argument: the witness commitment checks out IF it is serialized like a sha256d hash. See compute_witness_commitment

Reply:
I don't think they assume this even in the Bitcoin Core codebase yet ([see here](https://github.com/bitcoin/bitcoin/blob/a47319dada2f08d60b124395aedaf943dd63b2e9/src/validation.cpp#L3311)), could we use the slice as is and change when the commitment scheme is well-defined?

Reply:
done